doc/decisions/DR015-Strategy-for-methods-with-several-signature-variations.md

```diff
@@ -0,0 +1,179 @@
+[here](./DR003-ManagerInterface-method-naming.md). The salient points of
+the purposes of this discussion are:
+
+- There is no return value.
+- The `successCallback` is called for each element of the input vector
+  that is processed without error.
+- The `errorCallback` is called for any element of the input vector that
+  resulted in a processing error.
+
+Though this is flexible, and supports future design patterns, it can add
+significantly to the host implementation boilerplate for simple tasks
+such as resolving a single reference where any error is considered
+exceptional.
+
+We desire to add several host-facing conveniences that simplify common
+API interactions. This immediately precipitates the discussion as to how
+these should be integrated into the API.
+
+## Options considered
+
+### Option 1
+
+Use function overloading where possible.
+
+```c++
+void resolve(vector<T> inputs, F* successCallback, F* errorCallback);
+void resolve(vector<T> inputs, vector<variant<ValueType, Error>> &out);
+void resolve(vector<T> inputs, vector<ValueType> &out);
+void resolve(T input, variant<ValueType, Error> &out);
+void resolve(T input, ValueType &out);
+
+void entityVersions(vector<T> inputs, F* successCallback, F* errorCallback);
+void entityVersions(vector<T> inputs, vector<variant<ValueType, Error>> &out);
+void entityVersions(vector<T> inputs, vector<ValueType> &out);
+void entityVersions(T input, variant<ValueType, Error> &out);
+void entityVersions(T input, ValueType &out);
+```
+
+#### Pros
+
+- Low cognitive overhead to determine which high-level API method is
+  being invoked as the name contains no other information.
+- Avoids awkward compromises in determining method name mutations.
+
+#### Cons
+
+- Function overloading is not supported in many languages.
+- High cognitive overhead in understanding which variation is being used
+  as it relies on parameter naming and/or type visibility.
+- Forces out-params in C++ as overloading can't be based on return type.
+
+### Option 2
+
+Describe the variation as a continuation of the method name.
+
+```c++
+void resolve(vector<T> inputs, F* successCallback, F* errorCallback);
+vector<variant<ValueType, Error>> resolveToErrorOrResultVector(vector<T> inputs);
+vector<ValueType> resolveToResultVector(vector<T> inputs);
+variant<ValueType, Error> resolveOneToErrorOrResult(T input);
+ValueType resolveOneToResult(T input);
+
+void entityVersions(vector<T> inputs, F* successCallback, F* errorCallback);
+vector<variant<ValueType, Error>> entityVersionsAsErrorOrResultVector(vector<T> inputs);
+vector<ValueType> entityVersionsAsResultVector(vector<T> inputs);
+variant<ValueType, Error> entityVersionsForOneToErrorOrResult(T input);
+ValueType entityVersionsForOneToResult(T input);
+```
+
+#### Pros
+
+- Low cognitive overhead in understanding which variation is in use as
+  it is clearly described.
+
+#### Cons
+
+- Higher cognitive overhead to determine which high-level API method is
```

Comment:
This pattern might still have overloading problems since the variation is described, in the examples, in terms of return type